### PR TITLE
Adding action check to inform committers if they are targeting the wrong branch

### PR DESCRIPTION
**Pull request by jmherbst**
This fixes https://github.com/pi-hole/pi-hole/issues/3164

### By submitting this pull request, I confirm the following:
*please fill any appropriate checkboxes, e.g: [X]*

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [x] I have made only one major change in my proposed changes.
- [x] I have commented my proposed changes within the code.
- [x] I have tested my proposed changes, and have included unit tests where possible.
- [x] I am willing to help maintain this change if there are issues with it later.
- [x] I give this submission freely and claim no ownership.
- [x] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [x] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

Please make sure you [Sign Off](https://github.com/pi-hole/pi-hole/wiki/How-to-signoff-your-commits.) all commits. Pi-hole enforces the [DCO](https://github.com/pi-hole/pi-hole/wiki/Contributing-to-the-project).

---
### What does this PR aim to accomplish?:
I have added a CI check that runs a github action on every PR that targets a non-`development` branch and drops a comment on that PR informing the author that the base branch target is incorrect.

You can see an example of it in action over on https://github.com/jmherbst/pi-hole/pull/3
and an example of it not running when base branch is `development`: https://github.com/jmherbst/pi-hole/pull/2

### How does this PR accomplish the above?:
I've added a new github action workflow

### What documentation changes (if any) are needed to support this PR?:
None

---
* You must follow the template instructions. Failure to do so will result in your pull request being closed.
* Please respect that Pi-hole is developed by volunteers, who can only reply in their spare time.

### CODE REVIEW

File: .github/workflows/base-branch-check.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+        with:
+          message: 'Hello -- It appears you have opened this Pull Request to the wrong branch.
+
+                    ### Technical Requirements
+
+                    - Please submit Pull Requests to the **`development` branch only**.
+
+                    - Before Submitting your Pull Request, merge `development` with your new branch and fix any conflicts.
+                    (Make sure you don''t break anything in `development`!)'
```
Comment (dschaper) [inline .github/workflows/base-branch-check.yml:24]:
```suggestion
                    (Make sure you don't break anything in `development`!)'
```

File: .github/workflows/base-branch-check.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+name: Base Branch Checker
+
+on:
+  pull_request:
+    branches-ignore:
+      - development
```
Comment (dschaper) [inline .github/workflows/base-branch-check.yml:6]: Can this only include the `master` branch? PRs to other branches should be allowed without any warnings.

**Review by dschaper — DISMISSED** [review]

File: .github/workflows/base-branch-check.yml (hunk 2 above)
Comment (DL6ER) [inline .github/workflows/base-branch-check.yml:6]:
```suggestion
  pull_request_target:
    # By default, a workflow only runs when a pull_request_target's activity type is opened, synchronize, or reopened. 
    branches:
      - master # Run only on master branch
```

File: .github/workflows/base-branch-check.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+        with:
+          message: 'Hello -- It appears you have opened this Pull Request to the wrong branch.
+
+                    ### Technical Requirements
+
+                    - Please submit Pull Requests to the **`development` branch only**.
+
+                    - Before Submitting your Pull Request, merge `development` with your new branch and fix any conflicts.
+                    (Make sure you don't break anything in `development`!)'
```
Comment (DL6ER) [inline .github/workflows/base-branch-check.yml:24]:
```suggestion
        with:
          message: >
            Thank you very much for your contribution. An automated check has shown that you have opened your Pull Request to the `master` branch. As our project uses this branch for released code we ask you to submit any Pull Requests to the `development` branch instead. Note that, in most cases, you don't have to create a new pull request but can simply [change the base branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request).
```

File: .github/workflows/base-branch-check.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+name: Base Branch Checker
```
Comment (DL6ER) [inline .github/workflows/base-branch-check.yml:1]:
```suggestion
name: Pull Requests
```

File: .github/workflows/base-branch-check.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+name: Pull Requests
+
+on:
+  pull_request_target:
+    # By default, a workflow only runs when a pull_request_target's activity type is opened, synchronize, or reopened.
```
Comment (DL6ER) [inline .github/workflows/base-branch-check.yml:5]:
```suggestion
  # By default, a workflow only runs when a pull_request_target's activity type is opened, synchronize, or reopened. 
```

File: .github/workflows/base-branch-check.yml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+name: Pull Requests
+
+on:
+  pull_request_target:
+  # By default, a workflow only runs when a pull_request_target's activity type is opened, synchronize, or reopened.
```
Comment (DL6ER) [inline .github/workflows/base-branch-check.yml:5]:
```suggestion
  pull_request_target: # By default, a workflow only runs when a pull_request_target's activity type is opened, synchronize, or reopened. 
```

File: .github/workflows/base-branch-check.yml (hunk 2 above)
Comment (DL6ER) [reply .github/workflows/base-branch-check.yml]: I applied the change.

File: .github/workflows/base-branch-check.yml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+name: Pull Requests
+
+on:
+  pull_request_target: # By default, a workflow only runs when a pull_request_target's activity type is opened, synchronize, or reopened.
```
Comment (DL6ER) [inline .github/workflows/base-branch-check.yml:4]:
```suggestion
  pull_request_target: # By default, a workflow only runs when a pull_request_target's activity type is opened, synchronize, or reopened.
```

**Review by DL6ER — APPROVED** [review]

**Comment by DL6ER** [comment]
I squashed insignificant commits.

**Review by DL6ER — APPROVED** [review]

**Comment by yubiuser** [comment]
Needs to be ported to AdminLTE (change message target branch to `devel`) and maybe FTL

**Comment by DL6ER** [comment]
This is how the current proposal will look like:
Screenshot from 2021-12-21 13-50-02

**Review by yubiuser — DISMISSED** [review]

**Comment by yubiuser** [comment]
> Needs to be ported to AdminLTE

https://github.com/pi-hole/AdminLTE/pull/2013

File: .github/workflows/base-branch-check.yml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    branches:
+      - master # Run only on master branch
+
+jobs:
+  base_branch_check:
+    runs-on: ubuntu-latest
+    name: Base Branch Check
+    steps:
+      - name: Comment PR
```
Comment (PromoFaux) [inline .github/workflows/base-branch-check.yml:13]:
```suggestion
      - name: Get branch names
        id: branch-name
        uses: tj-actions/branch-names@v5.1
      - name: Comment PR
        if: steps.branch-name.outputs.head_ref_branch != 'development'
```

Sorry to throw a spanner in the works here, is there any value in adding this [action](https://github.com/tj-actions/branch-names) to the CI run here - which would prevent this comment from appearing when we merge `dev -> master` for release (I think)

Comment (PromoFaux) [reply .github/workflows/base-branch-check.yml:13]: I guess also we don't want it if branch starts with `release/` (but that should also be trivial to add as a conditional on the `if:` bit

File: .github/workflows/base-branch-check.yml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+        uses: thollander/actions-comment-pull-request@v1
+        with:
+          message: >
+            Thank you very much for your contribution. An automated check has shown that you have opened your Pull Request to the `master` branch. As our project uses this branch for released code we ask you to submit any Pull Requests to the `development` branch instead. Note that, in most cases, you don't have to create a new pull request but can simply [change the base branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request). After changing the base branch, ensure your branch does not contain any extra commits that aren't from you. If this is the case, please [rebase](https://anavarre.net/how-to-rebase-a-github-pull-request/) your branch.
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (DL6ER) [inline .github/workflows/base-branch-check.yml:18]:
```suggestion

  
name: PR Check

on:
  pull_request_target:
    # By default, a workflow only runs when a pull_request_target's activity type is opened, synchronize, or reopened. 
jobs:
  base_branch_check:
    runs-on: ubuntu-latest
    name: Base Branch
    steps:
      - name: Get branch names
        id: branch-name
        uses: tj-actions/branch-names@v5.1
      - name: Determine if PR is valid
        run: |
          echo "base_ref_branch = ${{ steps.branch-name.outputs.base_ref_branch }}"
          echo "head_ref_branch = ${{ steps.branch-name.outputs.head_ref_branch }}"
          if [[ "${{ steps.branch-name.outputs.base_ref_branch  == 'master' }}" = "true" ]]; then
            if [[ "${{ steps.branch-name.outputs.head_ref_branch  == 'development' }}" = "true" ]]; then
              echo "Valid PR: development -> master"
              echo "validPR=yes" >> $GITHUB_ENV
            elif [[ "${{ startsWith(steps.branch-name.outputs.head_ref_branch, 'release/') }}" = "true" ]]; then
              echo "Valid PR: release/... -> master"
              echo "validPR=yes" >> $GITHUB_ENV
            else
              echo "Invalid PR: ${{ steps.branch-name.outputs.head_ref_branch }} -> master"
              echo "validPR=no" >> $GITHUB_ENV
            fi
          else
            echo "Valid PR: ${{ steps.branch-name.outputs.head_ref_branch }} -> ${{ steps.branch-name.outputs.base_ref_branch }} "
            echo "validPR=yes" >> $GITHUB_ENV
          fi
          echo ::set-output name=value::${value}
      - name: Comment on invalid target
        if: env.validPR == 'no'
        uses: thollander/actions-comment-pull-request@v1
        with:
          message: >
            An automated check has shown that you have opened your Pull Request to the `master` branch. However, our project uses the `master` branch only for tested code.
            
            **Please submit a new Pull Request to the `development` branch.**
            
            You are the special type of person that makes Pi-hole better for us all! Thank you!
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      - name: Add label on invalid target
        if: env.validPR == 'no'
        uses: buildsville/add-remove-label@v1
        with:
          token: ${{secrets.GITHUB_TOKEN}}
          label: "Invalid target"
          type: add
      - name: Close invalid PR
        if: env.validPR == 'no'
        uses: superbrothers/close-pull-request@v3
      - name: Remove label on valid target # may have been added before
        if: env.validPR == 'yes'
        uses: buildsville/add-remove-label@v1
        with:
          token: ${{secrets.GITHUB_TOKEN}}
          label: "Invalid target"
          type: remove
```

**Comment by DL6ER** [comment]
Changed this to:
Screenshot from 2021-12-22 13-48-07

**Review by PromoFaux — CHANGES_REQUESTED** [review]
Blocking merge on this one for now - after some (failed) experiments on this PR : https://github.com/pi-hole/AdminLTE/pull/2014

It became clear that the `GITHUB_TOKEN` does not have enough permission to comment on PRs from forks. We would need to use a PAT - and I'm not comfortable enough with the (lack of) granularity with the permissions to use one for this purpose.

**Comment by DL6ER** [comment]
In addition to not being able to comment on PRs coming from forks, we've also realized that workflows don't even run by default for submissions from foreign repositories (including forks), rendering this entire endeavor useless.

**Pull request closed without merging**